File: practicas/practica2/src/main/java/deepspace/Damage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Representa el daño producido a una estación espacial por una naveenemiga cuando se pierde un combate
 */
public class Damage {
    private final static int NOUSED = -1;
    private int nShields;
    private int nWeapons;
    private ArrayList<WeaponType> weapons;

    Damage(int s, int w) {
        this.nShields = s;
        this.nWeapons = w;
        weapons = null;
    }
    
    Damage(ArrayList<WeaponType> wl, int s) {
        this.weapons = new ArrayList<>(wl);
        this.nShields = s;
        this.nWeapons = NOUSED;
    }
    
    Damage(Damage d) {
        this(d.getNWeapons(), d.getNShields());
    }
    
    DamageToUI getUIversion() {
        return new DamageToUI(this);
    }
    
    /**
     * @param w array de armas
     * @param t tipo de arma
     * @return devuelve el indice del primer elemento de w que sea de tipo t, -1 si no hay ninguno
     */
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t) {
        int index = -1;
        for (int i = 0; i < w.size(); i++) {
            if(w.get(i).getType() == t) {
                index = i;
                break;
            }
        }
        
        return index;
    }
    
    /**
     * Devuelve una versión ajustada del objeto a las colecciones de armas y potenciadores de escudos suministradas como parámetro.
     * @param w armas
     * @param s potenciadores
     * @return version ajustada del objeto
     */
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int shields = Integer.min(s.size(), nShields);
        if(weapons == null){
            //Caso tenemos contadores
            return new Damage(Integer.min(nWeapons, w.size()), shields);
        }
        
       ArrayList<Weapon> wAux = new ArrayList<>(w);
       ArrayList<WeaponType> toSet = new ArrayList<>();
       
       for(WeaponType element: weapons) {
          int index = -1;
          for(int i = 0; i < wAux.size(); i++){
              if(wAux.get(i).getType() == element){
                  index = i;
                  break;
              }
          }
          
          if(index != -1){
              toSet.add(element);
              wAux.remove(index);
          }
       }
        
        return new Damage(toSet, shields);
    }
    
    /**
     * intenta eliminar el tipo del arma pasada como parámetro de esa lista.
     * En otro caso simplemente decrementa en una unidad el contador de armas que deben ser eliminadas.
     * Ese contador no puedeser inferior a cero en ningún caso
     * @param w el arma
     */
    public void discardWeapon(Weapon w) {
        if(weapons == null){
            if(nWeapons > 0){
                nWeapons--;
            }
        } else {
            int index = weapons.indexOf(w.getType());
            if(index != -1){
                weapons.remove(index);
            }
        }
    }
    
    /**
     * Descarta un shieldBooster
     */
    public void discardShieldBooster() {
        nShields--;
        if(nShields < 0){
            nShields = 0;
        }
    }
    
    /**
     * @return true si el daño representado no tiene ningún efecto.
     * Esto quiere decir que no implica la pérdida de ningún tipo de accesorio (armas o potenciadores de escudo).
     */
    public boolean hasNoEffect() {
        if(weapons == null) {
            return (nWeapons == 0 && nShields == 0);
        }
        
        return (weapons.isEmpty() && nShields == 0);
    }

    public ArrayList<WeaponType> getWeapons() {
        if(weapons==null){
            return null;
        }
        return new ArrayList<>(weapons);
    }

    public int getNWeapons() {
        return nWeapons;
    }

    public int getNShields() {
        return nShields;
    }

    @Override
    public String toString() {
        if(weapons == null) {
            return "nShields: " + getNShields()+ "\nnWeapons: " + getNWeapons();
        }
        return "nShields: " + getNShields() + "\nWeapons: " + getWeapons().toString();
    }
    
    
    
}

File: practicas/practica3_rb/Damage.rb
#encoding: utf-8


#Clase que representa el daño producido a una estación espacial por una nave enemiga cuando se pierde un combate


module Deepspace

  class Damage

    NO_USE = -1

    private def initialize(nw, s, w)
      @nWeapons = nw
      @nShields = s
      @weapons = w
    end

    def self.newNumericWeapons(w, s)
      new(w, s, nil)
    end

    def self.newSpecificWeapons(w, s)
      new(NO_USE, s, w)
    end

      #Constructor de copia
    def self.newCopy(d)
      if(d.weapons!=nil)
        newd = Damage.newSpecificWeapons(d.weapons, d.nShields)
      else
        newd = Damage.newNumericWeapons(d.nWeapons, d.nShields)
      end
        return newd
    end

      #Consultores

    attr_reader :nShields

    attr_reader :nWeapons

    attr_reader :weapons

      #Alternativa 1 (solo borra el primero)
    def discardWeapon(w)
      indice = weapons.index(w.type)
        if indice != nil
          @weapons.delete_at(indice)
        else
          if nWeapons>0
            @nWeapons-=1
          else
            @nWeapons=0
          end
        end
    end

      #Alternativa 2 (borra todos)
    def discardWeapon(w)
      if weapons != nil
        @weapons.delete_if { |element| element == w.type }
      else
        if nWeapons > 0
          @nWeapons -= 1
        else
          @nWeapons = 0
        end
      end
    end

    def discardShieldBooster
      if nShields > 0
        @nShields-=1
      else
        @nShields = 0
      end
    end

    def hasNoEffect
      if weapons != nil
        if !weapons.empty
          return false
        end
      else
        if nWeapons > 0
          return false
        end
      end
        if nShields > 0
          return false
        end
        return true
    end

      private def arrayContainsType(w, t)
        indice = w.index(t)
          if(indice != nil)
            return indice
          else
            return -1
          end
      end



    def adjust(w, s)
      limit_nshields = [s.length, nShields].min

        if weapons==nil
          limit_nweapons = [w.length, nWeapons].min
            return Damage.newNumericWeapons(limit_nweapons, limit_nshields)

        else
          result = []
            w_aux = w.clone
            weapons.each do |element|
              indice = arrayContainsType(w_aux, element)
                

                if indice != -1
                  result.push(element)
                    w_aux.delete_at(indice)
                end
            end

            Damage.newSpecificWeapons(result, limit_nshields)

        end
    end



    def getUIversion
      DamageToUI.new(self)
    end

    def to_s
      if @weapons.nil?
        "nWeapons: " + @nWeapons.to_s + "\nnShields: " + @nShields.to_s + "\n"
      else
        "weapons: " + @weapons.to_s + "\nnShields: " + @nShields.to_s + "\n"
      end
    end

  end
end
